rust/arrow/src/compute/kernels/sort.rs

```diff
@@ -41,110 +41,123 @@ pub fn sort(values: &ArrayRef, options: Option<SortOptions>) -> Result<ArrayRef>
     take(values, &indices, None)
 }
 
+fn partition_nan<T: ArrowPrimitiveType>(
+    array: &ArrayRef,
+    v: Vec<u32>,
+) -> (Vec<u32>, Vec<u32>) {
+    // partition by nan for float types
+    if T::DATA_TYPE == DataType::Float32 {
+        // T::Native has no `is_nan` and thus we need to downcast
+        let array = array
+            .as_any()
+            .downcast_ref::<Float32Array>()
+            .expect("Unable to downcast array");
+        let has_nan = v.iter().any(|index| array.value(*index as usize).is_nan());
```

Comment:
A minor comment is that you might be able  to make the code smaller (avoid scanning the array twice) if we don't bother to check for `has_nan`-- and just always do `v.into_iter().partition(....)`.  The rationale is if `has_nan` is false, you had to scan the entire array anyways.

I have not tested my statement, and it is based on assumptions of the compiler being clever about optimizing `into_iter()`

Reply:
But I see this is just refactored from below, so no need to change